feat(observability): trace Database methods

odeke-em opened the pull request:
Adds trace spans for Database methods, as well as tests
for methods:

* getSession
* getSnapshot
* run
* runStream
* runTransaction

tracing of other methods shall come in follow-up PRs.

Updates #2079
Built from PR #2087
Updates #2114

odeke-em committed:
feat(observability): trace Database methods

Adds trace spans for Database methods, as well as tests
for methods:

* getSession
* getSnapshot
* run
* runStream
* runTransaction

tracing of other methods shall come in follow-up PRs.

Updates #2079
Built from PR #2087
Updates #2114

odeke-em commented:
@surbhigarg92 kindly please re-review!

surbhigarg92 commented on test/spanner.ts:
```diff
@@ -5046,8 +5048,10 @@ describe('Spanner with mock server', () => {
         'Creating 25 sessions',
-        'Acquiring session',
-        'Waiting for a session to become available',
+        // Since we added new spans in the Database methods
```
Please remove the commented code.

odeke-em replied:
Alright.

surbhigarg92 commented on observability-test/spanner.ts:
```diff
@@ -0,0 +1,367 @@
+            expectedSpanNames,
+            `span names mismatch:\n\tGot:  ${actualSpanNames}\n\tWant: ${expectedSpanNames}`
+          );
+
```
Add checks for events as well in E2E tests

odeke-em replied:
Done, thank you @surbhigarg92!

surbhigarg92 commented on observability-test/database.ts:
```diff
@@ -0,0 +1,1285 @@
+      }
+    });
+
+    it('should return CommitResponse on successful write using Callback', done => {
```
Looks like a copy paste. Remove all these test where we are not adding check for spans

odeke-em replied:
Done, thank you!

odeke-em commented on test/database.ts:
```diff
@@ -1992,7 +1992,7 @@ describe('Database', () => {
         })
         .on('end', () => {
-          assert.strictEqual(endStub.callCount, 1);
+          assert.strictEqual(endStub.callCount, 2);
```
Odd commit, sorry.

odeke-em commented:
Hello @surbhigarg92 kindly please take a look. Thank you!

odeke-em commented:
@surbhigarg92 @harshachinta kindly please help trigger the bot runs. Thank you.

surbhigarg92 commented on test/database.ts:
```diff
@@ -840,9 +840,12 @@ describe('Database', () => {
+        // depending on the version of Node.js, the error might be either of:
+        //  * Cannot read properties of null (reading 'proto')
+        //  * Cannot read property 'proto' of null
+        (err as grpc.ServiceError).message.includes('Cannot read propert');
```
instead of this do something like 
const errorMessage = (err as grpc.ServiceError).message;
errorMessage.includes("Cannot read properties of null (reading 'proto')") || 
  errorMessage.includes("Cannot read property 'proto' of null")

odeke-em replied:
Awesome and thank you @surbhigarg92!

surbhigarg92 commented on test/database.ts:
```diff
@@ -2961,7 +2964,9 @@ describe('Database', () => {
-      assert.strictEqual(runFn, fakeRunFn);
+      // Given that we've wrapped the transaction runner with observability
+      // tracing, directly comparing values runFn and fakeRunFn.
+      // assert.strictEqual(runFn, fakeRunFn);
```
is it possible to add a check on some properties of these functions, instead of removing this check ?

odeke-em replied:
Sadly not, unless we are perform a traversal by utility examination but I found that to be overkill and not worth it TBH

surbhigarg92 reviewed (approved):
Please fix the 2 comments before merging

odeke-em commented:
@surbhigarg92 thank you very much for the reviews! I've addressed all feedback and gotten the PR ready for merge by squashing commits into one. Kindly help me re-run the bots.
